modules/ephemeris/src/main/scala/gem/horizons/HorizonsNameQuery.scala

```diff
@@ -0,0 +1,235 @@
+        }
+
+      // Single result with form: JPL/HORIZONS      90377 Sedna (2003 VB12)     2015-Dec-31 11:40:21
+      def case1: ParsedAsteroids =
+        """  +\d+ ([^(]+)\s+\((.+?)\)  """.r.findFirstMatchIn(header).map { m =>
+          List(Resolution(EphemerisKey.AsteroidNew(m.group(2)) : EphemerisKey.Asteroid, m.group(1)))
+        }.toRight("Could not match '90377 Sedna (2003 VB12)' header pattern.")
+
+      // Single result with form: JPL/HORIZONS      4 Vesta     2015-Dec-31 11:40:21
+      def case2: ParsedAsteroids =
+        """  +(\d+) ([^(]+?)  """.r.findFirstMatchIn(header).map { m =>
+          List(Resolution(EphemerisKey.AsteroidOld(m.group(1).toInt) : EphemerisKey.Asteroid, m.group(2)))
+        }.toRight("Could not match '4 Vesta' header pattern.")
+
+      // Single result with form: JPL/HORIZONS    (2016 GB222)    2016-Apr-20 15:22:36
+      def case3: ParsedAsteroids =
+        """  +\((.+?)\)  """.r.findFirstMatchIn(header).map { m =>
+          List(Resolution(EphemerisKey.AsteroidNew(m.group(1)) : EphemerisKey.Asteroid, m.group(1)))
+        }.toRight("Could not match '(2016 GB222)' header pattern.")
+
+      // Single result with form: JPL/HORIZONS        418993 (2009 MS9)            2016-Sep-07 18:23:54
+      def case4: ParsedAsteroids =
+        """  +\d+\s+\((.+?)\)  """.r.findFirstMatchIn(header).map { m =>
+          List(Resolution(EphemerisKey.AsteroidNew(m.group(1)) : EphemerisKey.Asteroid, m.group(1)))
+        }.toRight("Could not match '418993 (2009 MS9)' header pattern.")
+
+      // First one that works!
+      case0 orElse
+      case1 orElse
+      case2 orElse
+      case3 orElse
+      case4 orElse "Could not parse the header line as an asteroid".asLeft
+    }
+    // scalastyle:on method.length
+
+    def parseMajorBodies(header: String, tail: List[String]): ParsedMajorBodies = {
+
+      // Common case is that we have many results, or none.
+      def case0: ParsedMajorBodies =
+        parseMany[Resolution[EphemerisKey.MajorBody]](header, tail, """Multiple major-bodies match string""".r) { offs =>
+          (offs.lift(0), offs.lift(1)).mapN {
+            case ((ors, ore), (ons, one)) => { row =>
+              val rec  = row.substring(ors, ore).trim.toInt
+              val name = row.substring(ons, one).trim
+              Resolution(EphemerisKey.MajorBody(rec.toInt), name)
+            }
+          }
+        }.map(_.filterNot(_.a.num < 0)) // filter out spacecraft
+
+      // Single result with form:  Revised: Aug 11, 2015       Charon / (Pluto)     901
+      def case1: ParsedMajorBodies =
+        """  +(.*?) / \((.+?)\)  +(\d+) *$""".r.findFirstMatchIn(header).map { m =>
+          List(Resolution(EphemerisKey.MajorBody(m.group(3).toInt), m.group(1)))
+        }.toRight("Could not match 'Charon / (Pluto)     901' header pattern.")
+
+      // First one that works, otherwise Nil because it falls through to small-body search
+      case0 orElse
+      case1 orElse Nil.asRight
+    }
+  }
```

Comment:
I guess the Horizons service doesn't offer a more orthodox format like JSON. Is the Horizons format formally documented somewhere? Why not use `atto` in this case?

Reply:
Unfortunately Horizons doesn't really offer formatted data in its responses and we end up having to parse information that was meant to be read by a human.  In this case the various cases are tricky so I opted to just keep the regex parsing solution that we've had in place for a few years instead of trying to re-implement it.